pkg/storage/parse.go

```diff
@@ -7,32 +7,30 @@ import (
 	"github.com/pingcap/kvproto/pkg/backup"
 )
 
-// BackendOption further configures the storage backend not expressed by the
+// BackendOptions further configures the storage backend not expressed by the
 // storage URL.
-type BackendOption interface {
-	OptionName() string
-	applyOnNoop(noop *backup.Noop, name string) error
-	applyOnLocal(local *backup.Local, name string) error
-	applyOnS3(s3 *backup.S3, name string) error
+type BackendOptions struct {
+	S3 S3BackendOptions `json:"s3" toml:"s3"`
 }
 
-type backendOption struct{}
-
-func (backendOption) applyOnNoop(_ *backup.Noop, name string) error {
-	return errors.Errorf("option '%s' is not applicable to noop storage", name)
-}
+func (options *BackendOptions) Validate(storageScheme string) error {
+	optionsNotApplicableToStorage := func(option string) error {
+		return errors.Errorf("options '%s.*' are not applicable to %s storage", option, storageScheme)
+	}
 
-func (backendOption) applyOnLocal(_ *backup.Local, name string) error {
-	return errors.Errorf("option '%s' is not applicable to local storage", name)
-}
+	if !options.S3.isEmpty() && storageScheme != "s3" {
+		return optionsNotApplicableToStorage("s3")
+	}
+	if err := options.S3.validate(); err != nil {
+		return err
+	}
 
-func (backendOption) applyOnS3(_ *backup.Noop, name string) error {
-	return errors.Errorf("option '%s' is not applicable to s3 storage", name)
+	return nil
```

Comment:
This is not straightforward, can we make it like below:
```
switch storageScheme{
case s3:
  if options.S3.isEmpty(){
    return error
  }
  validate()
```
If the above is accepted, we may just migrate the logic to the blocks from L48.